File: src/backtracking/cpp/knight_tour/src/knight_tour.cpp
#include "knight_tour.h"
#include <algorithm>
#include <array>

#include <unordered_map>
#include <utility>

struct PairHash {
  template <class T1, class T2>
  std::size_t operator()(const std::pair<T1, T2> &p) const {
    auto h1 = std::hash<T1>{}(p.first);
    auto h2 = std::hash<T2>{}(p.second);
    return h1 ^ h2;
  }
};

std::vector<std::vector<int>>
transform_tour(const std::vector<std::pair<int, int>> &tour, int n) {
  // create a map from squares to move numbers
  std::unordered_map<std::pair<int, int>, int, PairHash> move_map;
  for (int i = 0; i < tour.size(); ++i) {
    move_map[tour[i]] = i + 1;
  }

  // create a list of moves by looking up the move numbers in the map
  std::vector<std::vector<int>> moves;
  for (int row = 0; row < n; ++row) {
    std::vector<int> row_moves;
    for (int col = 0; col < n; ++col) {
      row_moves.push_back(move_map[{row, col}]);
    }
    moves.push_back(row_moves);
  }
  return moves;
}

void find_tours(std::vector<std::vector<int>> &board, int row, int col,
                int move_count, std::vector<std::pair<int, int>> &tour,
                std::vector<std::vector<std::pair<int, int>>> &tours) {
  // base case: all squares on the board have been visited
  if (move_count == board.size() * board[0].size()) {
    tours.push_back(tour);
    return;
  }

  // try all possible moves from the current position
  static const std::array<std::pair<int, int>, 8> possible_moves = {{
      {2, 1},
      {2, -1},
      {-2, 1},
      {-2, -1},
      {1, 2},
      {1, -2},
      {-1, 2},
      {-1, -2},
  }};
  for (const auto &[row_offset, col_offset] : possible_moves) {
    int new_row = row + row_offset;
    int new_col = col + col_offset;
    if (new_row >= 0 && new_row < board.size() && new_col >= 0 &&
        new_col < board[0].size() && board[new_row][new_col] == -1) {
      board[new_row][new_col] = move_count;
      tour.emplace_back(new_row, new_col);
      find_tours(board, new_row, new_col, move_count + 1, tour, tours);
      tour.pop_back();
      board[new_row][new_col] = -1;
    }
  }
}

std::vector<std::vector<std::vector<int>>> knight_tour(int n) {
  // initialize the board and the tours list
  std::vector<std::vector<int>> board(n, std::vector<int>(n, -1));
  board[0][0] = 0;
  std::vector<std::pair<int, int>> tour = {{0, 0}};
  std::vector<std::vector<std::pair<int, int>>> tours;
  find_tours(board, 0, 0, 1, tour, tours);

  // transform the tours into a list of move lists
  std::vector<std::vector<std::vector<int>>> tours_list;
  for (const auto &tour : tours) {
    tours_list.push_back(transform_tour(tour, n));
  }
  return tours_list;
}

File: src/backtracking/python/knight_tour/src/knight_tour.py
from typing import List, Tuple, Optional, Dict


def transform_tour(tour: List[Tuple[int, int]], n: int) -> List[List[int]]:
    """
    Transform a list of tour positions into a grid of move numbers.

    Args:
        tour: List of (row, col) positions representing the knight's path.
        n: Size of the board.

    Returns:
        An n x n grid where each cell contains the move number (1-indexed).
    """
    # create a map from squares to move numbers
    move_map: Dict[Tuple[int, int], int] = {
        square: i + 1 for i, square in enumerate(tour)
    }

    # create a list of moves by looking up the move numbers in the map
    moves = [[move_map[(row, col)] for col in range(n)] for row in range(n)]
    return moves


def find_tours(
    board: List[List[Optional[int]]],
    row: int,
    col: int,
    move_count: int,
    tour: List[Tuple[int, int]],
    tours: List[List[Tuple[int, int]]],
) -> None:
    """
    Recursively find all knight's tour solutions using backtracking.

    Args:
        board: The current state of the board.
        row: Current row position.
        col: Current column position.
        move_count: Number of moves made so far.
        tour: Current tour path.
        tours: List to store all complete tours.
    """
    # base case: all squares on the board have been visited
    if move_count == len(board) * len(board[0]):
        tours.append(tour.copy())
        return

    # try all possible moves from the current position
    for row_offset, col_offset in [
        (2, 1),
        (2, -1),
        (-2, 1),
        (-2, -1),
        (1, 2),
        (1, -2),
        (-1, 2),
        (-1, -2),
    ]:
        new_row, new_col = row + row_offset, col + col_offset
        if (
            0 <= new_row < len(board)
            and 0 <= new_col < len(board[0])
            and board[new_row][new_col] is None
        ):
            board[new_row][new_col] = move_count
            tour.append((new_row, new_col))
            find_tours(board, new_row, new_col, move_count + 1, tour, tours)
            tour.pop()
            board[new_row][new_col] = None


def knight_tour(n: int) -> List[List[List[int]]]:
    """
    Find all knight's tour solutions for an n x n board.

    A knight's tour is a sequence of moves where the knight visits
    every square exactly once.

    Args:
        n: Size of the board.

    Returns:
        List of all valid tours, where each tour is an n x n grid
        showing the order of moves.
    """
    # initialize the board and the tours list
    board: List[List[Optional[int]]] = [[None] * n for _ in range(n)]
    board[0][0] = 0
    tour: List[Tuple[int, int]] = [(0, 0)]
    tours: List[List[Tuple[int, int]]] = []
    find_tours(board, 0, 0, 1, tour, tours)
    tours_result = [transform_tour(tour, n) for tour in tours]
    return tours_result
